先將 orchid_private_set.zip 和 orchid_public_set.zip
解壓縮至 ./AI_CUP_2022_Orchid_TEAM_511/AI_CUP_2022_test/orchid_set/內做存放


**File Structure**

orchid_set

    ├── orchid_public_set
    │     ├──*.jpg
    │     │
    │     │  ...
    │     │
    │     ├──*.jpg
    │      
    └── orchid_private_set
          ├──*.jpg
          │
          │  ...
          │
          ├──*.jpg



**1.在Colab上掛載Google Drive**

In [ ]:
from google.colab import drive
drive.mount("/content/drive")

Mounted at /content/drive


**2.進入到文件所在的目錄**

In [ ]:
import os
path = "/content/drive/My Drive/AI_CUP_2022_Orchid_TEAM_511"
os.chdir(path)
os.listdir(path)

['Final_Test.py',
 'Ans',
 'AI_CUP_2022_test',
 'Datasets',
 'pytorch-image-models',
 'STN',
 '=0.6']

**3.需安裝套件:**
`pip install timm==0.6.2.dev0`


In [ ]:
pip install timm==0.6.2.dev0

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 497 kB 5.6 MB/s 


**4.執行 Final_Test.py 即可輸出答案，答案會儲存在(./Ans)**

`!python Final_Test.py`


*＃注意
本隊伍在實驗階段時有發現到Pillow的版本若非當初訓練時的版本(8.3.1)，在輸出結果上會有些許誤差。* 


In [ ]:
!python Final_Test.py

0/6,0a2xqs6vrl.jpg
6/5


**5.如何執行 train.py**

**5-1 進入pytorch-image-models 目錄下**


In [ ]:
%cd pytorch-image-models

/content/drive/My Drive/AI_CUP_2022_Orchid_TEAM_511/pytorch-image-models


**5-2 訓練 swin large**



--experiment 為儲存的檔案夾

`!CUDA_LAUNCH_BLOCKING=1 python train.py ../Datasets/training --train-split train --val-split val --model swin_large_patch4_window12_384_in22k --pretrained --num-classes 219 --mean 0.5075337404927281 0.45864544276917535 0.4169235386212412 --std 0.2125643051799512 0.2385082849964861 0.22386483801695406 -b 4 -vb 4 --opt adamw --weight-decay 0.01 --layer-decay 0.65 --sched cosine --lr 0.0001 --lr-cycle-limit 1 --warmup-lr 1e-5 --min-lr 1e-5 --epochs 300 --warmup-epochs 5 --color-jitter 0.5 --reprob 0.5 --recount 1 --remode pixel --cutmix 0.4 --train-interpolation bicubic --drop-path 0.1 -j 4 --save-images --output output --experiment 0`

In [ ]:
!CUDA_LAUNCH_BLOCKING=1 python train.py ../Datasets/training --train-split train --val-split val --model swin_large_patch4_window12_384_in22k --pretrained --num-classes 219 --mean 0.5075337404927281 0.45864544276917535 0.4169235386212412 --std 0.2125643051799512 0.2385082849964861 0.22386483801695406 -b 4 -vb 4 --opt adamw --weight-decay 0.01 --layer-decay 0.65 --sched cosine --lr 0.0001 --lr-cycle-limit 1 --warmup-lr 1e-5 --min-lr 1e-5 --epochs 300 --warmup-epochs 5 --color-jitter 0.5 --reprob 0.5 --recount 1 --remode pixel --cutmix 0.4 --train-interpolation bicubic --drop-path 0.1 -j 4 --save-images --output output --experiment 0

Training with a single process on 1 GPUs.
/usr/local/lib/python3.7/dist-packages/torch/functional.py:568: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at  ../aten/src/ATen/native/TensorShape.cpp:2228.)
  return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
Loading pretrained weights from url (https://github.com/SwinTransformer/storage/releases/download/v1.0.0/swin_large_patch4_window12_384_22k.pth)
Model swin_large_patch4_window12_384_in22k created, param count:195535119
Data processing configuration for current model + dataset:
	input_size: (3, 384, 384)
	interpolation: bicubic
	mean: (0.5075337404927281, 0.45864544276917535, 0.4169235386212412)
	std: (0.2125643051799512, 0.2385082849964861, 0.22386483801695406)
	crop_pct: 1.0
parameter groups: 
{
  "layer_0_decay": {
    "lr_scale": 2.1029740616282293e-05,
    "weight_decay": 0.01,
    "param_names": [
      "patch_embed.proj.weight"
    

**5-2-1 Fine tune swin large**



`!CUDA_LAUNCH_BLOCKING=1 python train.py ../Datasets/training --train-split train --val-split val --model swin_large_patch4_window12_384_in22k --initial-checkpoint ./model/0/model_best.pth.tar --num-classes 219 --mean 0.5075337404927281 0.45864544276917535 0.4169235386212412 --std 0.2125643051799512 0.2385082849964861 0.22386483801695406 -b 2 -vb 2 --opt adamw --weight-decay 0.01 --sched step --lr 0.00001 --decay-rate 1.0 --no-aug --epochs 100 --warmup-epochs 0 --color-jitter 0.0 --reprob 0.0 --recount 1 --remode pixel --cutmix 0.0 --train-interpolation bicubic --drop-path 0.0 --log-interval 30 -j 4 --save-images --output output --experiment 0_no_aug_no_weight_decay`

**5-3-1 訓練Swin v2 base (A-fold)**

`!CUDA_LAUNCH_BLOCKING=1 python train.py ../Datasets/training --train-split train --val-split val --model swinv2_base_window12to24_192to384_22kft1k --pretrained --num-classes 219 --img-size 384 --mean 0.5075337404927281 0.45864544276917535 0.4169235386212412 --std 0.2125643051799512 0.2385082849964861 0.22386483801695406 -b 2 -vb 2 --opt adamw --weight-decay 0.01 --layer-decay 0.65 --sched cosine --lr 0.0001 --lr-cycle-limit 1 --warmup-lr 1e-5 --min-lr 1e-5 --epochs 300 --warmup-epochs 5 --color-jitter 0.5 --reprob 0.5 --recount 1 --remode pixel --cutmix 0.4 --train-interpolation bicubic --drop-path 0.1 -j 4 --save-images --output output --experiment 0531_0310_swinv2_base_window12to24_192to384_22kft1k_official_aug_v0_new`

**5-3-2 訓練Swin v2 base (B-fold)**

`!CUDA_LAUNCH_BLOCKING=1 python train.py ../Datasets/AI_CUP_fold_complement --train-split train --val-split val --model swinv2_base_window12to24_192to384_22kft1k --pretrained --num-classes 219 --mean 0.45625534556274666 0.4220624936173144 0.3649616601198825 --std 0.2143212828816861 0.2210437745632625 0.2062174242104951 -b 2 -vb 2 --opt adamw --weight-decay 0.01 --layer-decay 0.65 --sched cosine --lr 0.0001 --lr-cycle-limit 2 --warmup-lr 1e-5 --min-lr 1e-5 --epochs 300 --warmup-epochs 5 --color-jitter 0.5 --reprob 0.5 --recount 1 --remode pixel --cutmix 0.4 --train-interpolation bicubic --drop-path 0.1 -j 4 --save-images --output output --experiment 0603_0131_swinv2_base_window12to24_192to384_22kft1k_complement_final_bacbbone_new --mixup-off-epoch 300`